packages/grafana-ui/src/components/uPlot/Plot.tsx

```diff
@@ -59,22 +53,38 @@ export const UPlotChart: React.FC<PlotProps> = props => {
     };
   }, [plotInstance]);
 
+  useLayoutEffect(() => {
+    plotData.current = preparePlotData(props.data);
+  }, [props.data]);
+
   // Decides if plot should update data or re-initialise
-  useEffect(() => {
-    if (!currentConfig) {
+  useLayoutEffect(() => {
+    // Make sure everything is ready before proceeding
+    if (!currentConfig || !plotData.current) {
+      return;
+    }
+
+    // Do nothing if there is data vs series config mismatch. This may happen when the data was updated and made this
+    // effect fire before the config update triggered the effect.
+    if (currentConfig.series.length !== plotData.current.length) {
```

Comment:
@mckn this is the actual bug fix... Not super happy about this flow yet, think would be good if we could sit down at some point and think about improving the uPlot init/data update logic